File: Week_01/python/21_merge_two_sorted_lists.py
from sys import maxsize


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def mergeTwoLists(self, l1: ListNode, l2: ListNode) -> ListNode:
        if not l1: return l2
        if not l2: return l1

        current = ListNode(-maxsize)
        dummy = current

        while l1 or l2:
            if not l1:
                current.next = l2
                break
            if not l2:
                current.next = l1
                break

            if l1.val > l2.val:
                current.next = l2
                l2 = l2.next
                current = current.next
            else:
                current.next = l1
                l1 = l1.next
                current = current.next

        return dummy.next

File: Week_01/java/21_merge_two_sorted_lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        if( l1 == null) return l2;
        if( l2 == null) return l1;

        ListNode current = new ListNode(-2^31 + 1);

        ListNode dummy = current;

        while(l1 != null || l2 != null) {

            if( l1 == null) {
                current.next = l2;
                break;
            }

            if( l2 == null) {
                current.next = l1;
                break;
            }

            if( l1.val >= l2.val){
                current.next = l2;
                l2 = l2.next;
                current = current.next;
            } else {
                current.next = l1;
                l1 = l1.next;
                current = current.next;
            }
        }
        return dummy.next;
    }

}
